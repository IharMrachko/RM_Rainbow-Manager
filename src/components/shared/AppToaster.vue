<template>
  <teleport to="body">
    <div class="toast-container">
      <transition-group name="toast" tag="div">
        <div v-for="toast in toasts" :key="toast.id" class="toast" :class="toast.type">
          <span class="toast-icon">
            <font-awesome-icon :icon="iconClass(toast.type)" />
          </span>
          <span class="toast-message">{{ toast.message }}</span>
          <div class="toast-progress" :style="{ animationDuration: toast.duration + 'ms' }"></div>
          <!-- Кнопка удаления -->
          <div class="toast-delete" @click="deleteToast(toast.id)">
            <font-awesome-icon :icon="['fas', 'times-circle']" />
          </div>
        </div>
      </transition-group>
    </div>
  </teleport>
</template>

<script lang="ts">
import { computed, defineComponent } from 'vue';
import { useStore } from 'vuex';
import { ToastType } from '@/store/modules/toast';

export default defineComponent({
  setup() {
    const store = useStore();
    const toasts = computed(() => store.state.toast.toasts);
    const deleteToast = (id: number) => {
      store.dispatch('toast/deleteToast', {
        id,
      });
    };
    const iconClass = (type: ToastType) => {
      switch (type) {
        case 'success':
          return ['fas', 'check-circle'];
        case 'error':
          return ['fas', 'times-circle'];
        case 'info':
          return ['fas', 'info-circle'];
        case 'warning':
          return ['fas', 'exclamation-triangle'];
        default:
          return ['fas', 'times-circle'];
      }
    };
    return {
      toasts,
      iconClass,
      deleteToast,
    };
  },
});
</script>

<style scoped lang="scss">
$toast-colors: (
  success: #57b846,
  error: #f44336,
  info: #2196f3,
  warning: #ff9800,
);

.toast-container {
  position: fixed;
  top: 1rem;
  right: 1rem;
  z-index: 9999;
  display: flex;
  flex-direction: column;
  gap: 0.75rem;
  pointer-events: none;
}

.toast {
  min-width: 280px;
  max-width: 360px;
  padding: 0.75rem 1rem;
  border-radius: 8px;
  font-size: 0.95rem;
  font-weight: 500;
  box-shadow: 3px 6px 18px rgba(0, 0, 0, 0.45);
  background-color: rgba(0, 0, 0, 0.8);
  position: relative;
  overflow: hidden;
  pointer-events: auto;
  display: flex;
  align-items: center;
  gap: 0.5rem;
  margin-bottom: 0.5rem;

  &.success {
    .toast-icon {
      color: map-get($toast-colors, success);
    }
    .toast-progress {
      background-color: map-get($toast-colors, success);
    }
  }
  &.error {
    .toast-icon {
      color: map-get($toast-colors, error);
    }
    .toast-progress {
      background-color: map-get($toast-colors, error);
    }
  }
  &.info {
    .toast-icon {
      color: map-get($toast-colors, info);
    }
    .toast-progress {
      background-color: map-get($toast-colors, info);
    }
  }
  &.warning {
    .toast-icon {
      color: map-get($toast-colors, warning);
    }
    .toast-progress {
      background-color: map-get($toast-colors, warning);
    }
  }

  .toast-icon {
    font-size: 1.2rem;
    flex-shrink: 0;
    opacity: 0.9;
  }

  .toast-message {
    flex: 1;
    color: #e6e6e6;
  }

  .toast-progress {
    position: absolute;
    bottom: 0;
    left: 0;
    height: 4px;
    background-color: rgba(255, 255, 255, 0.7);
    animation-name: progress;
    animation-timing-function: linear;
    animation-fill-mode: forwards;
  }

  .toast-delete {
    position: absolute;
    top: 6px;
    right: 6px;
    width: 20px;
    height: 20px;
    display: flex;
    align-items: center;
    justify-content: center;
    cursor: pointer;
    border-radius: 50%;
    pointer-events: auto; // чтобы клик работал

    svg {
      color: #fff;
    }
  }
}

@keyframes progress {
  from {
    width: 100%;
  }
  to {
    width: 0;
  }
}

/* Анимация появления/исчезновения */
.toast-enter-from {
  transform: translateX(100%);
  opacity: 0;
}
.toast-enter-to {
  transform: translateX(0);
  opacity: 1;
}
.toast-enter-active {
  transition: all 0.4s ease-out;
}

.toast-leave-from {
  opacity: 1;
  transform: translateY(0);
}
.toast-leave-to {
  opacity: 0;
  transform: translateY(-20px);
}
.toast-leave-active {
  transition: all 0.4s ease-in;
}
</style>
