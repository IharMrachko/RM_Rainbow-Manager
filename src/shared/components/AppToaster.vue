<template>
  <teleport to="body">
    <div class="toast-container">
      <transition-group name="toast" tag="div">
        <div
          v-for="toast in toasts"
          id="toastTest"
          :key="toast.id"
          class="toast"
          :class="toast.severity"
        >
          <span class="toast-icon">
            <font-awesome-icon :icon="iconClass(toast.severity)" />
          </span>
          <span class="toast-message">{{ t(toast.message) }}</span>
          <div class="toast-progress" :style="{ animationDuration: toast.duration + 'ms' }"></div>
          <div class="toast-delete" @click="deleteToast(toast.id)">
            <font-awesome-icon :icon="['fas', 'times-circle']" />
          </div>
        </div>
      </transition-group>
    </div>
  </teleport>
</template>

<script lang="ts">
import { computed, defineComponent } from 'vue';
import { useStore } from 'vuex';
import { Severity } from '@/types/color.type';
import { useI18n } from 'vue-i18n';

export default defineComponent({
  setup() {
    const { t } = useI18n();
    const store = useStore();
    const toasts = computed(() => store.getters['toast/getToasts']);
    const deleteToast = (id: number) => {
      store.dispatch('toast/deleteToast', {
        id,
      });
    };
    const iconClass = (severity: Severity) => {
      switch (severity) {
        case 'success':
          return ['fas', 'check-circle'];
        case 'error':
          return ['fas', 'times-circle'];
        case 'info':
          return ['fas', 'info-circle'];
        case 'warning':
          return ['fas', 'exclamation-triangle'];
        default:
          return ['fas', 'times-circle'];
      }
    };
    return {
      toasts,
      iconClass,
      deleteToast,
      t,
    };
  },
});
</script>

<style scoped lang="scss">
@import '@/styles/style';
.toast-container {
  position: fixed;
  top: 1rem;
  right: 1rem;
  z-index: 9999;
  display: flex;
  flex-direction: column;
  gap: 0.75rem;
  pointer-events: none;
}

.toast {
  min-width: 280px;
  max-width: 360px;
  padding: 0.75rem 1rem;
  border-radius: 8px;
  font-size: 0.95rem;
  font-weight: 500;
  box-shadow: 3px 6px 18px rgba(0, 0, 0, 0.45);
  background-color: rgba(0, 0, 0, 0.8);
  position: relative;
  overflow: hidden;
  pointer-events: auto;
  display: flex;
  align-items: center;
  gap: 0.5rem;
  margin-bottom: 0.5rem;

  &.success {
    .toast-icon {
      color: map-get($type-colors, success);
    }
    .toast-progress {
      background-color: map-get($type-colors, success);
    }
  }
  &.error {
    .toast-icon {
      color: map-get($type-colors, error);
    }
    .toast-progress {
      background-color: map-get($type-colors, error);
    }
  }
  &.info {
    .toast-icon {
      color: map-get($type-colors, info);
    }
    .toast-progress {
      background-color: map-get($type-colors, info);
    }
  }
  &.warning {
    .toast-icon {
      color: map-get($type-colors, warning);
    }
    .toast-progress {
      background-color: map-get($type-colors, warning);
    }
  }

  .toast-icon {
    font-size: 1.2rem;
    flex-shrink: 0;
    opacity: 0.9;
  }

  .toast-message {
    flex: 1;
    color: #e6e6e6;
  }

  .toast-progress {
    position: absolute;
    bottom: 0;
    left: 0;
    height: 4px;
    background-color: rgba(255, 255, 255, 0.7);
    animation-name: progress;
    animation-timing-function: linear;
    animation-fill-mode: forwards;
  }

  .toast-delete {
    position: absolute;
    top: 6px;
    right: 6px;
    width: 20px;
    height: 20px;
    display: flex;
    align-items: center;
    justify-content: center;
    cursor: pointer;
    border-radius: 50%;
    pointer-events: auto; // чтобы клик работал

    svg {
      color: #fff;
    }
  }
}

@keyframes progress {
  from {
    width: 100%;
  }
  to {
    width: 0;
  }
}

/* Анимация появления/исчезновения */
.toast-enter-from {
  transform: translateX(100%);
  opacity: 0;
}
.toast-enter-to {
  transform: translateX(0);
  opacity: 1;
}
.toast-enter-active {
  transition: all 0.4s ease-out;
}

.toast-leave-from {
  opacity: 1;
  transform: translateY(0);
}
.toast-leave-to {
  opacity: 0;
  transform: translateY(-20px);
}
.toast-leave-active {
  transition: all 0.4s ease-in;
}
</style>
