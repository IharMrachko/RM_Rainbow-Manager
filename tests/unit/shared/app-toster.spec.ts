import { createStore } from 'vuex';
import { flushPromises, mount } from '@vue/test-utils';
import AppToaster from '@/components/shared/AppToaster.vue';
import { nextTick } from 'vue';

describe('AppToaster', () => {
  let store: any;
  let actions: any;
  let state: any;

  beforeEach(() => {
    actions = {
      deleteToast: jest.fn(),
    };
    state = {
      toast: {
        toasts: [
          { id: 1, message: 'Success message', type: 'success', duration: 3000 },
          { id: 2, message: 'Error message', type: 'error', duration: 5000 },
        ],
      },
    };
    store = createStore({
      modules: {
        toast: {
          namespaced: true,
          state: state.toast,
          actions,
        },
      },
    });
  });
  afterEach(() => {
    document.body.innerHTML = '';
  });
  it("renders correctly and matches snapshot'", () => {
    const wrapper = mount(AppToaster, {
      global: { plugins: [store] },
      attachTo: document.body,
    });
    expect(document.body.innerHTML).toMatchSnapshot();
    wrapper.unmount();
  });
  it('renders toasts from store', async () => {
    const wrapper = mount(AppToaster, {
      global: { plugins: [store] },
      attachTo: document.body,
    });

    await nextTick();
    await flushPromises();
    const toastEls = document.body.querySelectorAll('#toastTest');

    expect(toastEls).toHaveLength(2);
    expect(toastEls[0].textContent).toContain('Success message');
    expect(toastEls[1].textContent).toContain('Error message');
    wrapper.unmount();
  });

  it('applies correct class for toast type', () => {
    const wrapper = mount(AppToaster, {
      global: { plugins: [store] },
      attachTo: document.body,
    });
    const toastEls = document.body.querySelectorAll('.toast');
    expect(toastEls[0].classList).toContain('success');
    expect(toastEls[1].classList).toContain('error');
    wrapper.unmount();
  });

  it('sets correct animationDuration for progress bar', () => {
    const wrapper = mount(AppToaster, {
      global: { plugins: [store] },
      attachTo: document.body,
    });

    const progressEls = document.body.querySelectorAll('.toast-progress');
    expect(progressEls[0].getAttribute('style')).toContain('3000ms');
    expect(progressEls[1].getAttribute('style')).toContain('5000ms');
    wrapper.unmount();
  });

  it('calls deleteToast action when delete button clicked', async () => {
    const wrapper = mount(AppToaster, {
      global: { plugins: [store] },
      attachTo: document.body, // чтобы teleport рендерился в body
    });

    // Ищем кнопки удаления в body
    const deleteButtons = document.body.querySelectorAll<HTMLDivElement>('.toast-delete');

    expect(deleteButtons).toHaveLength(2); // убеждаемся, что нашли обе

    // Кликаем по первой
    deleteButtons[0].dispatchEvent(new MouseEvent('click', { bubbles: true }));

    // Ждём обновления Vue
    await wrapper.vm.$nextTick();

    expect(actions.deleteToast).toHaveBeenCalledWith(expect.any(Object), { id: 1 });

    wrapper.unmount(); // чистим body после теста
  });

  it('returns correct icon for each type', () => {
    const wrapper = mount(AppToaster, {
      global: { plugins: [store] },
    });
    const vm = wrapper.vm as any;
    expect(vm.iconClass('success')).toEqual(['fas', 'check-circle']);
    expect(vm.iconClass('error')).toEqual(['fas', 'times-circle']);
    expect(vm.iconClass('info')).toEqual(['fas', 'info-circle']);
    expect(vm.iconClass('warning')).toEqual(['fas', 'exclamation-triangle']);
    expect(vm.iconClass('unknown')).toEqual(['fas', 'times-circle']);
  });
});
