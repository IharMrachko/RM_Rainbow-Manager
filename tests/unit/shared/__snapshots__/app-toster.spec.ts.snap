// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppToaster renders correctly and matches snapshot' 1`] = `
<div data-v-app=""><!--teleport start--><!--teleport end--></div>
<div class="toast-container"><transition-group-stub name="toast" tag="div" appear="false" persisted="false" css="true">
    <div id="toastTest" class="toast success"><span class="toast-icon"><font-awesome-icon icon="fas,check-circle"></font-awesome-icon></span><span class="toast-message">Success message</span>
      <div class="toast-progress" style="animation-duration: 3000ms;"></div>
      <div class="toast-delete"><font-awesome-icon icon="fas,times-circle"></font-awesome-icon></div>
    </div>
    <div id="toastTest" class="toast error"><span class="toast-icon"><font-awesome-icon icon="fas,times-circle"></font-awesome-icon></span><span class="toast-message">Error message</span>
      <div class="toast-progress" style="animation-duration: 5000ms;"></div>
      <div class="toast-delete"><font-awesome-icon icon="fas,times-circle"></font-awesome-icon></div>
    </div>
  </transition-group-stub></div>
`;
